File: app/src/main/java/com/example/asthmaapplication/main/mainpage/MainFragment.java
package com.example.asthmaapplication.main.mainpage;

import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentTransaction;
import androidx.lifecycle.ViewModelProvider;

import android.os.Bundle;

import androidx.annotation.Nullable;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.example.asthmaapplication.R;
import com.example.asthmaapplication.databinding.FragmentMainBinding;
import com.example.asthmaapplication.main.common.BaseFragment;
import com.example.asthmaapplication.main.common.SnackBarMessage;
import com.example.asthmaapplication.main.homepage.HomePageFragment;
import com.example.asthmaapplication.main.mainpage.subpages.LearningBaseFragment;
import com.example.asthmaapplication.main.mainpage.subpages.PatientsFragment;
import com.example.asthmaapplication.main.mainpage.subpages.QuizzesExamsFragment;
import com.example.asthmaapplication.main.mainpage.subpages.ReviewsFragment;
import com.example.asthmaapplication.main.utils.UIUtils;

import javax.inject.Inject;

import dagger.hilt.android.AndroidEntryPoint;
import io.reactivex.annotations.NonNull;

@AndroidEntryPoint
public class MainFragment extends BaseFragment {
    MainViewModel viewModel;
    FragmentMainBinding binding;
    FragmentManager manager;
    FragmentTransaction transaction;

    @Inject
    public MainFragment() {

    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        viewModel = new ViewModelProvider(this).get(MainViewModel.class);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @javax.annotation.Nullable ViewGroup container, @javax.annotation.Nullable Bundle savedInstanceState) {
        binding = FragmentMainBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@Nullable View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        getPreferences();
        setActionBarTitle();
        validateUserAccess();

        manager = getFragmentManager();
        transaction = manager.beginTransaction();

        binding.learningCard.cardImage.setImageResource(R.drawable.ic_lung_icon);
        binding.learningCard.cardTitle.setText(R.string.learning_card);
        binding.learningCard.getRoot().setOnClickListener(v -> launchLearningPage());

        binding.patientsCard.cardImage.setImageResource(R.drawable.ic_patient_icon);
        binding.patientsCard.cardTitle.setText(R.string.patient_card);

        binding.quizCard.cardImage.setImageResource(R.drawable.ic_quiz_icon);
        binding.quizCard.cardTitle.setText(R.string.quiz_card);

        binding.reviewCard.cardImage.setImageResource(R.drawable.ic_review_icon);
        binding.reviewCard.cardTitle.setText(R.string.review_card);

        UIUtils.addUnderlineFlag(binding.actionLogout);

        binding.actionLogout.setOnClickListener(userLogout -> {
            viewModel.logOut();
            getActivity().finish();
            clearUserName();
            launchHomePage();
        });

    }

    @Override
    public void onResume() {
        super.onResume();
    }

    public void setActionBarTitle() {
        if (getUserName().isEmpty()) {
            setReadingPreferences(false);
            binding.actionLogout.setText(getString(R.string.redirect_home));
            setActionBarTitle(getString(R.string.welcome_user, getString(R.string.guest_user)));
        } else {
            binding.actionLogout.setText(getString(R.string.log_out));
            setActionBarTitle(getString(R.string.welcome_user, getUserName()));
        }
    }


    public void validateUserAccess() {
        if (getUserName().isEmpty() && !getHasUserRead()) {
            binding.patientsCard.getRoot().setOnClickListener(v -> showSnackBar(new SnackBarMessage(getString(R.string.guest_user_access_warning))));
            binding.quizCard.getRoot().setOnClickListener(v -> showSnackBar(new SnackBarMessage(getString(R.string.guest_user_access_warning))));
            binding.reviewCard.getRoot().setOnClickListener(v -> showSnackBar(new SnackBarMessage(getString(R.string.guest_user_access_warning))));
        } else {
            binding.patientsCard.getRoot().setOnClickListener(v -> launchPatientsCard());
            binding.quizCard.getRoot().setOnClickListener(v -> launchQuizPage());
            binding.reviewCard.getRoot().setOnClickListener(v -> launchReviewsPage());
        }
    }

    public void launchLearningPage() {
        LearningBaseFragment fragment = new LearningBaseFragment();
        transaction
                .addToBackStack(null)
                .replace(R.id.fragment_container, fragment)
                .commit();

    }

    public void launchPatientsCard() {
        PatientsFragment fragment = new PatientsFragment();
        transaction
                .addToBackStack(null)
                .replace(R.id.fragment_container, fragment)
                .commit();
    }

    public void launchReviewsPage() {
        ReviewsFragment fragment = new ReviewsFragment();
        transaction
                .addToBackStack(null)
                .replace(R.id.fragment_container, fragment)
                .commit();
    }

    public void launchQuizPage() {
        QuizzesExamsFragment fragment = new QuizzesExamsFragment();
        transaction
                .addToBackStack(null)
                .replace(R.id.fragment_container, fragment)
                .commit();
    }

    public void launchHomePage() {
        HomePageFragment fragment = new HomePageFragment();
        transaction
                .addToBackStack(null)
                .replace(R.id.fragment_container, fragment)
                .commit();
    }

    @Override
    public View getRoot() {
        return binding.getRoot();
    }
}

File: app/src/main/java/com/example/asthmaapplication/main/mainpage/MainFragment.kt
package com.example.asthmaapplication.main.mainpage

import dagger.hilt.android.AndroidEntryPoint
import com.example.asthmaapplication.main.common.BaseFragment
import android.os.Bundle
import android.view.*
import androidx.core.view.MenuHost
import androidx.core.view.MenuProvider
import androidx.fragment.app.viewModels
import androidx.lifecycle.Lifecycle
import com.example.asthmaapplication.R
import com.example.asthmaapplication.databinding.FragmentMainBinding
import com.example.asthmaapplication.main.common.Constants
import com.example.asthmaapplication.main.utils.UIUtils
import com.example.asthmaapplication.main.common.SnackBarMessage
import com.example.asthmaapplication.main.mainpage.myprofile.MyProfileFragment
import com.example.asthmaapplication.main.mainpage.subpages.learningbasepages.LearningBaseFragment
import com.example.asthmaapplication.main.mainpage.subpages.learningbasepages.PatientsFragment
import com.example.asthmaapplication.main.mainpage.subpages.learningbasepages.ReviewsFragment
import com.example.asthmaapplication.main.mainpage.subpages.learningbasepages.QuizzesExamsFragment
import com.example.asthmaapplication.main.utils.ActivityUtils

@AndroidEntryPoint
class MainFragment: BaseFragment() {
    private lateinit var _binding: FragmentMainBinding

    private val _viewModel: MainViewModel by viewModels()

    private var _loginOption: String? = null
    private var _userName: String? = null

    private val _myProfileFragment = MyProfileFragment()

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        _binding = FragmentMainBinding.inflate(inflater, container, false)
        return _binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setupUi()
        setupListeners()
        setupMenu()
    }

    override fun onResume() {
        super.onResume()
        _loginOption = _viewModel.userType
        setActionBarTitle()
        validateUserAccess()
    }

    private fun setupListeners() {
        _binding.actionLogout.setOnClickListener {
            _viewModel.logOut()
            activity?.finish()
        }
    }

    private fun setupUi() {
        UIUtils.addUnderlineFlag(_binding.actionLogout)
        _binding.learningCard.cardImage.setImageResource(R.drawable.ic_lung_normal)
        _binding.learningCard.cardTitle.setText(R.string.learning_card)
        _binding.patientsCard.cardImage.setImageResource(R.drawable.ic_patient_icon)
        _binding.patientsCard.cardTitle.setText(R.string.patient_card)
        _binding.quizCard.cardImage.setImageResource(R.drawable.ic_quiz_icon)
        _binding.quizCard.cardTitle.setText(R.string.quiz_card)
        _binding.reviewCard.cardImage.setImageResource(R.drawable.ic_review_icon)
        _binding.reviewCard.cardTitle.setText(R.string.review_card)
    }

    private fun setupMenu() {
        (requireActivity() as MenuHost).addMenuProvider(object : MenuProvider {
            override fun onCreateMenu(menu: Menu, menuInflater: MenuInflater) {
                menuInflater.inflate(R.menu.my_profile_menu, menu)
            }

            override fun onMenuItemSelected(item: MenuItem): Boolean {
                if (item.itemId == R.id.action_my_profile) {
                    ActivityUtils.addFragmentWithBackStack(
                        parentFragmentManager,
                        _myProfileFragment,
                        R.id.fragment_container,
                        "my_profile"
                    )
                    return true
                }
                return false
            }
        }, viewLifecycleOwner, Lifecycle.State.RESUMED)

    }

    private fun setActionBarTitle() {
        _userName = _viewModel.userName

        if (_userName.isNullOrEmpty()) {
            _binding.actionLogout.text = getString(R.string.redirect_home)
            _binding.mainTitleDescription.text = getString(R.string.guest_user_tip)
            setActionBarTitle(getString(R.string.welcome_user, getString(R.string.guest_user)))
        } else {
            _binding.mainTitleDescription.text = getText(R.string.main_page_title)
            _binding.actionLogout.text = getString(R.string.log_out)
            setActionBarTitle(getString(R.string.welcome_user, _userName))
        }
    }

    private fun validateUserAccess() {
        when (_loginOption) {
            Constants.STUDENT_OPTION -> {
                setStudentAccess()
            }
            Constants.PATIENT_OPTION -> {
                setPatientAccess()
            }
            else -> {
                setDefaultAccess()
            }
        }
    }

    private fun setStudentAccess() {
        _binding.learningCard.root.setOnClickListener { launchLearningPage() }
        _binding.patientsCard.root.setOnClickListener { launchPatientsCard() }
        _binding.quizCard.root.setOnClickListener { launchQuizPage() }
        _binding.reviewCard.root.setOnClickListener { launchReviewsPage() }
    }

    private fun setPatientAccess() {
        _binding.quizCard.cardHolder.setBackgroundColor(resources.getColor(R.color.gray))
        _binding.reviewCard.cardHolder.setBackgroundColor(resources.getColor(R.color.gray))
        _binding.learningCard.root.setOnClickListener { launchLearningPage() }
        _binding.patientsCard.root.setOnClickListener { launchPatientsCard() }
        _binding.reviewCard.root.setOnClickListener { setAccessWarnings(getString(R.string.patient_access_warning)) }
        _binding.quizCard.root.setOnClickListener { setAccessWarnings(getString(R.string.patient_access_warning)) }
    }

    private fun setDefaultAccess() {
        _binding.learningCard.root.setOnClickListener { launchLearningPage() }
        _binding.patientsCard.cardHolder.setBackgroundColor(resources.getColor(R.color.gray))
        _binding.quizCard.cardHolder.setBackgroundColor(resources.getColor(R.color.gray))
        _binding.reviewCard.cardHolder.setBackgroundColor(resources.getColor(R.color.gray))
        _binding.patientsCard.root.setOnClickListener { setAccessWarnings(getString(R.string.guest_user_access_warning)) }
        _binding.quizCard.root.setOnClickListener { setAccessWarnings(getString(R.string.guest_user_access_warning)) }
        _binding.reviewCard.root.setOnClickListener { setAccessWarnings(getString(R.string.guest_user_access_warning)) }
    }

    private fun setAccessWarnings(warningMessage: String) {
        showSnackBar(SnackBarMessage(warningMessage))
    }

    private fun launchLearningPage() {
        val fragment = LearningBaseFragment()
        ActivityUtils.addFragmentWithBackStack(
            parentFragmentManager,
            fragment,
            R.id.fragment_container,
            "learning_page"
        )
    }

    private fun launchPatientsCard() {
        val fragment = PatientsFragment()
        ActivityUtils.addFragmentWithBackStack(
            parentFragmentManager,
            fragment,
            R.id.fragment_container,
            "learning_page"
        )
    }

    private fun launchReviewsPage() {
        val fragment = ReviewsFragment()
        ActivityUtils.addFragmentWithBackStack(
            parentFragmentManager,
            fragment,
            R.id.fragment_container,
            "learning_page"
        )
    }

    private fun launchQuizPage() {
        val fragment = QuizzesExamsFragment()
        ActivityUtils.addFragmentWithBackStack(
            parentFragmentManager,
            fragment,
            R.id.fragment_container,
            "learning_page"
        )
    }

    override fun getRoot(): View {
        return _binding.root
    }
}